src/App.tsx

```diff
@@ -388,11 +478,44 @@ async function fetchAndVerifyAddress({
   proof.balance = numberToHex(proof.balance);
 
   const proofStateRoot = toHexString(keccak256(toBuffer(proof.accountProof[0])));
-  const verified =
+  let verified =
     stateRoot === proofStateRoot &&
-    proof.storageHash === externalAddressStorageHash &&
-    proof.codeHash === externalAddressCodeHash &&
+    (proof.codeHash !== externalAddressCodeHash || proof.storageHash === externalAddressStorageHash) &&
     (await stateManager.verifyProof(proof));
+  let tokens = [];
+
+  for (const contractName of Object.keys(erc20Contracts)) {
+    const {contractAddress, balanceMappingIndex} = erc20Contracts[contractName];
+    const balanceSlot = web3.utils.keccak256(
+      defaultAbiCoder.encode(["address", "uint"], [address, balanceMappingIndex])
+    );
+    const contractProof = await web3.eth.getProof(contractAddress, [balanceSlot], executionPayload.blockNumber);
+    if (contractProof.codeHash === externalAddressCodeHash) {
+      throw Error(`No contract deployed at ${contractAddress} for ${contractName}`);
+    }
+    const contractProofStateRoot = toHexString(keccak256(toBuffer(contractProof.accountProof[0])));
+    // Verify the proof, web3 converts nonce and balance into number strings, however
+    // ethereumjs verify proof requires them in the original hex format
+    contractProof.nonce = numberToHex(contractProof.nonce);
+    contractProof.balance = numberToHex(contractProof.balance);
+
+    verified =
+      verified &&
+      stateRoot === contractProofStateRoot &&
+      contractProof.storageProof[0]?.key === balanceSlot &&
+      (await stateManager.verifyProof(contractProof));
```

Comment:
This is async for local cyrpto ops or it does a remote call?

Reply:
async for local, it does some Trie traversal/walk, i think to load the proof